File: platform/polyglot/rag_ui_component_ingest_config/mjs/src/schema.mjs
/**
 * @fileoverview Zod v3 validation schemas for rag_ui_component_ingest_config.
 *
 * All schemas use camelCase keys to match the MJS internal representation.
 * The Python counterpart uses snake_case (Pydantic); when exchanging data over
 * the wire use the `toDict()` methods that emit snake_case.
 */

import { z } from 'zod';

// ---------------------------------------------------------------------------
// BaseIngestConfigSchema
// ---------------------------------------------------------------------------

/**
 * Validates all shared infrastructure configuration fields.
 * Mirrors Python's `BaseIngestConfigSchema` (Pydantic).
 */
export const BaseIngestConfigSchema = z
  .object({
    // Paths
    datasetRoot: z.string().min(1),
    persistRoot: z.string().min(1),

    // Embeddings
    embeddingsModelName: z.string().min(1),

    // Chunking
    chunkSize: z.number().int().positive(),
    chunkOverlap: z.number().int().nonnegative(),

    // Vector backend
    vectorBackend: z.enum(['chroma', 'elasticsearch']),

    // Elasticsearch
    elasticsearchHost: z.string().min(1),
    elasticsearchPort: z.number().int().positive(),
    elasticsearchScheme: z.string().min(1),

    // Redis
    redisHost: z.string().min(1),
    redisPort: z.number().int().positive(),

    // LLM
    llmProvider: z.string().min(1),
    openaiModel: z.string().min(1),
    anthropicModel: z.string().min(1),
    geminiModel: z.string().min(1),

    // Search / retrieval
    hybridAlpha: z.number().min(0).max(1),
    scoreThreshold: z.number().min(0),
    rerankerEnabled: z.boolean(),
    rerankerModel: z.string().min(1),
    retrieveN: z.number().int().positive(),
    topK: z.number().int().positive(),

    // Database
    postgresEnabled: z.boolean(),
  })
  .refine(
    (data) => data.chunkOverlap < data.chunkSize,
    (data) => ({
      message: `chunkOverlap (${data.chunkOverlap}) must be less than chunkSize (${data.chunkSize})`,
      path: ['chunkOverlap'],
    }),
  );

// ---------------------------------------------------------------------------
// LibraryConfigSchema
// ---------------------------------------------------------------------------

/**
 * Validates a single library configuration entry.
 * Mirrors Python's `LibraryConfigSchema`.
 */
export const LibraryConfigSchema = z.object({
  // Identity — required
  name: z.string().min(1),
  slug: z
    .string()
    .regex(
      /^[a-z0-9][a-z0-9-]*[a-z0-9]$/,
      'slug must be URL-safe lowercase with at least 2 characters',
    ),

  // Optional overrides
  version: z.string().optional(),
  sourceDirectory: z.string().optional(),
  persistDirectory: z.string().optional(),
  examplesDirectory: z.string().optional(),
  elasticsearchIndex: z.string().optional(),
  componentPathSegment: z.string().default('components'),

  importPackages: z.array(z.string()).default([]),
  chunkSize: z.number().int().positive().optional(),
  chunkOverlap: z.number().int().nonnegative().optional(),

  fileExtensions: z.array(z.string()).optional(),
  ignoredDirectories: z.array(z.string()).optional(),

  enabled: z.boolean().default(true),
});

// ---------------------------------------------------------------------------
// DocumentMetadataSchema
// ---------------------------------------------------------------------------

/**
 * Source type discriminator — the canonical set of values.
 */
export const SOURCE_TYPE_ENUM = z.enum([
  'component',
  'story',
  'doc',
  'style',
  'type',
  'test',
  'config',
]);

/**
 * Validates the metadata attached to a single document chunk.
 * Mirrors Python's `DocumentMetadataSchema`.
 */
export const DocumentMetadataSchema = z.object({
  // Identity
  library: z.string().min(1),
  libraryVersion: z.string().optional(),
  component: z.string().optional(),

  // Source
  fileName: z.string().min(1),
  filePath: z.string().min(1),
  sourceType: SOURCE_TYPE_ENUM,
  language: z.string(),

  // Lineage
  contentHash: z.string(),
  chunkIndex: z.number().int().nonnegative(),
  totalChunks: z.number().int().positive(),
  ingestedAt: z.string(),

  // Navigation
  heading: z.string().optional(),
  exportName: z.string().optional(),
});

File: platform/polyglot/rag_ui_component_ingest_config/py/rag_ui_component_ingest_config/schema.py
"""Pydantic v2 validation schemas for rag_ui_component_ingest_config."""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field, field_validator, model_validator


class BaseIngestConfigSchema(BaseModel):
    """Validated view of all shared infrastructure configuration fields."""

    # Paths
    dataset_root: str = Field(..., min_length=1)
    persist_root: str = Field(..., min_length=1)

    # Embeddings
    embeddings_model_name: str = Field(..., min_length=1)

    # Chunking
    chunk_size: int = Field(..., gt=0)
    chunk_overlap: int = Field(..., ge=0)

    # Vector backend
    vector_backend: Literal["chroma", "elasticsearch"]

    # Elasticsearch
    elasticsearch_host: str = Field(..., min_length=1)
    elasticsearch_port: int = Field(..., gt=0)
    elasticsearch_scheme: str = Field(..., min_length=1)

    # Redis
    redis_host: str = Field(..., min_length=1)
    redis_port: int = Field(..., gt=0)

    # LLM
    llm_provider: str = Field(..., min_length=1)
    openai_model: str = Field(..., min_length=1)
    anthropic_model: str = Field(..., min_length=1)
    gemini_model: str = Field(..., min_length=1)

    # Search / retrieval
    hybrid_alpha: float = Field(..., ge=0.0, le=1.0)
    score_threshold: float = Field(..., ge=0.0)
    reranker_enabled: bool
    reranker_model: str = Field(..., min_length=1)
    retrieve_n: int = Field(..., gt=0)
    top_k: int = Field(..., gt=0)

    # Database
    postgres_enabled: bool

    @model_validator(mode="after")
    def chunk_overlap_less_than_chunk_size(self) -> "BaseIngestConfigSchema":
        if self.chunk_overlap >= self.chunk_size:
            raise ValueError(
                f"chunk_overlap ({self.chunk_overlap}) must be less than "
                f"chunk_size ({self.chunk_size})"
            )
        return self

    model_config = {"extra": "forbid"}


class LibraryConfigSchema(BaseModel):
    """Validated view of a single library configuration entry."""

    # Identity — required
    name: str = Field(..., min_length=1)
    slug: str = Field(
        ...,
        pattern=r"^[a-z0-9][a-z0-9-]*[a-z0-9]$",
        description="URL-safe lowercase identifier",
    )

    # Optional overrides
    version: Optional[str] = None
    source_directory: Optional[str] = None
    persist_directory: Optional[str] = None
    examples_directory: Optional[str] = None
    elasticsearch_index: Optional[str] = None
    component_path_segment: str = "components"

    import_packages: list[str] = Field(default_factory=list)
    chunk_size: Optional[int] = Field(default=None, gt=0)
    chunk_overlap: Optional[int] = Field(default=None, ge=0)

    file_extensions: Optional[list[str]] = None
    ignored_directories: Optional[list[str]] = None

    enabled: bool = True

    @field_validator("slug")
    @classmethod
    def slug_min_two_chars(cls, v: str) -> str:
        # The regex requires at least two characters (start + end char class).
        # Single-char slugs like "a" do not satisfy `^[a-z0-9][a-z0-9-]*[a-z0-9]$`
        # because the `*` between the two anchored char classes needs 0+ middle
        # chars — that means the slug must be at least 2 chars.
        return v

    model_config = {"extra": "forbid"}


class DocumentMetadataSchema(BaseModel):
    """Validated view of a single document chunk's metadata."""

    # Identity
    library: str = Field(..., min_length=1)
    library_version: Optional[str] = None
    component: Optional[str] = None

    # Source
    file_name: str = Field(..., min_length=1)
    file_path: str = Field(..., min_length=1)
    source_type: Literal["component", "story", "doc", "style", "type", "test", "config"]
    language: str

    # Lineage
    content_hash: str
    chunk_index: int = Field(..., ge=0)
    total_chunks: int = Field(..., ge=1)
    ingested_at: str

    # Navigation
    heading: Optional[str] = None
    export_name: Optional[str] = None

    model_config = {"extra": "allow"}
